de.variantsync.core/src/main/java/de/variantsync/core/ast/AST.java

```diff
@@ -0,0 +1,183 @@
+	public static transient final String NEXT_SEPARATOR = "\u2502 ";
+	public static transient final String NEXT_ACT_SEPARATOR = "\u251C\u2500 ";
+	public static transient final String LAST_SEPARATOR = "\u2514\u2500 ";
+
+	public AST(UUID id, G type, V value) {
+		this.id = id;
+		this.type = type;
+		this.value = value;
+		this.subtrees = new ArrayList<>();
+	}
+
+	public AST(G type, V value) {
+		this(UUID.randomUUID(), type, value);
+	}
+
+	/**
+	 * Empty AST is forbidden at the moment.
+	 */
+	private AST() {
+
+	}
+
+	public UUID getId() {
+		return id;
+	}
+
+	public V getValue() {
+		return value;
+	}
+
+	public G getType() {
+		return type;
+	}
+
+	/**
+	 * @return the subtrees as an unmodifiable List
+	 */
+	public List<AST<G, V>> getSubtrees() {
+		return Collections.unmodifiableList(subtrees);
+	}
+
+	public int getMaxDepth() {
+		if (subtrees.size() == 0) {
+			return 0;
+		}
+
+		int maxDepth = 0;
+		for (final AST<G, V> node : subtrees) {
+			maxDepth = Math.max(node.getMaxDepth(), maxDepth);
+		}
+		return ++maxDepth;
+	}
+
+	/**
+	 * This method calls the addChild method for each element of the given list. ASTs which would be rejected by the addChild method are skipped.
+	 *
+	 * @param toAdd List of AST which should be added as subtrees
+	 * @return true if all items where successfully added
+	 */
+	public boolean addChildren(List<AST<G, V>> toAdd) {
+		boolean out = true;
+		if (toAdd != null) {
+			for (final AST<G, V> elem : toAdd) {
+				if (!addChild(elem)) {
+					out = false;
+				}
+			}
+		}
+		return out;
+	}
+
+	/**
+	 * This method adds an element to the AST by checking its validity through the the Grammar G and assuring that the element is not null.
+	 *
+	 * @param toAdd Single AST which should be added as subtree
+	 * @return true if the item was successfully added
+	 */
+	public boolean addChild(AST<G, V> toAdd) {
+		if ((toAdd != null) && type.isValidChild(toAdd.type)) {
+			return subtrees.add(toAdd);
+		}
+		return false;
+	}
+
+	public int size() {
+		int tmpSize = 1;
+		for (final AST<G, V> act : subtrees) {
+			tmpSize += act.size();
+		}
+		return tmpSize;
+	}
+
+	/**
+	 * This method returns only the most significant bits of the UUID, the type, value and subtree size as a String of the actual AST.
+	 *
+	 * @return UUID, Type, Value, subtree size as String
+	 */
+	@Override
+	public String toString() {
+		return String.format("[ Id: %d, Type: %s, Value: %s, Subtree-size: %d ]", id.getMostSignificantBits(), type.toString(), value.toString(),
+				subtrees.size());
+	}
+
+	/**
+	 * This recursive method prints for each tree element the Grammar type, the Value and (for the sake of readability) only the most significant bits of the
+	 * UUID. It returns the whole AST as human readable tree.
+	 *
+	 * @return AST as readable String
+	 */
+	public String printTree() {
+		final StringBuilder result = new StringBuilder();
+		if (value == null) {
+			return result.toString();
```

Comment:
Duplicate code with the very last line of this method. You could just remove this branch.